src/attrib.d

```diff
@@ -611,23 +624,91 @@ public:
 extern (C++) final class AlignDeclaration : AttribDeclaration
 {
 public:
-    uint salign;
+    Expression ealign;
+    structalign_t salign = STRUCTALIGN_DEFAULT;
 
-    extern (D) this(uint sa, Dsymbols* decl)
+    extern (D) this(Loc loc, Expression ealign, Dsymbols* decl)
     {
         super(decl);
-        salign = sa;
+        this.loc = loc;
+        this.ealign = ealign;
     }
 
     override Dsymbol syntaxCopy(Dsymbol s)
     {
         assert(!s);
-        return new AlignDeclaration(salign, Dsymbol.arraySyntaxCopy(decl));
+        return new AlignDeclaration(loc,
+            ealign.syntaxCopy(), Dsymbol.arraySyntaxCopy(decl));
     }
 
     override Scope* newScope(Scope* sc)
     {
-        return createNewScope(sc, sc.stc, sc.linkage, sc.protection, sc.explicitProtection, this.salign, sc.inlining);
+        return createNewScope(sc, sc.stc, sc.linkage,
+            sc.protection, sc.explicitProtection, this,
+            sc.inlining);
+    }
+
+    override void setScope(Scope* sc)
+    {
+        //printf("AlignDeclaration::setScope() %p\n", this);
+        if (ealign && decl)
+            Dsymbol.setScope(sc); // for forward reference
+        return AttribDeclaration.setScope(sc);
+    }
+
+    override void semantic2(Scope* sc)
+    {
+        getAlignment();
+        super.semantic2(sc);
+    }
+
+    structalign_t getAlignment()
+    {
+        if (!ealign)
+            return STRUCTALIGN_DEFAULT;
+
+        if (auto sc = _scope)
+        {
+            _scope = null;
+
+            sc = sc.startCTFE();
+            ealign = ealign.semantic(sc);
+            ealign = resolveProperties(sc, ealign);
+            sc = sc.endCTFE();
+
+            auto showError()
+            {
+                .error(loc, "positive integer expected, not %s", ealign.toChars());
+                return STRUCTALIGN_DEFAULT;
+            }
+
+            if (ealign.op == TOKerror)
+                return STRUCTALIGN_DEFAULT;
+            if (!ealign.type)
+                return showError();
+            Type tb = ealign.type.toBasetype();
+            if (!tb.isintegral())
+                return showError();
+            if (tb.ty == Tchar || tb.ty == Twchar || tb.ty == Tdchar || tb.ty == Tbool)
+                return showError();
+
+            ealign = ealign.ctfeInterpret();
+            if (ealign.op == TOKerror)
+                return STRUCTALIGN_DEFAULT;
+
+            auto n = ealign.toInteger();
+            if (!tb.isunsigned() && n > long.max)
```

Comment:
Please read my repply. Currently alignment value greater than long.max is silently accepted. Your proposal would be legitimiate, but it's unrelated with the new feature implementation. Shouldn't it go to separated bugfix PR?